src/ucp/stream/stream_recv.c

```diff
@@ -56,6 +74,24 @@ ucp_stream_recv_data_nb_internal(ucp_ep_ext_stream_t *ep_stream)
     return rdesc;
 }
 
+static UCS_F_ALWAYS_INLINE ucp_recv_desc_t *
+ucp_stream_recv_data_nb_internal(ucp_ep_ext_stream_t *ep_stream, int dequeue)
+{
+    ucp_recv_desc_t *rdesc;
+
+    if (ucs_queue_is_empty(&ep_stream->data)) {
+        return NULL;
+    }
+
+    rdesc = dequeue ? ucp_stream_rdesc_dequeue(ep_stream) :
```

Comment:
ucp_stream_rdesc_dequeue() also checks if the queue is empty, need to avoid checking twice

Reply:
This function return NULL if the queue is empty but  `ucp_stream_rdesc_dequeue()` checks if the queue is empty **after** dequeuing. If it was last element then remove `ep` from another queue if it does not have received data. For example, user had posted `recv_nb` or called `recv_data_nb` before worker poll.